src/apps/relay/windows/ServiceInstaller.cpp

```diff
@@ -0,0 +1,162 @@
+    printf("OpenSCManager failed (%lu)\n", GetLastError());
+    return -2;
+  }
+
+  // Install the service into SCM by calling CreateService
+  schService = CreateService(schSCManager,              // SCManager database
+                             pszServiceName,            // Name of service
+                             pszDisplayName,            // Name to display
+                             SERVICE_QUERY_STATUS,      // Desired access
+                             SERVICE_WIN32_OWN_PROCESS, // Service type
+                             dwStartType,               // Service start type
+                             SERVICE_ERROR_NORMAL,      // Error control type
+                             szPath,                    // Service's binary
+                             NULL,                      // No load ordering group
+                             NULL,                      // No tag identifier
+                             pszDependencies,           // Dependencies
+                             pszAccount,                // Service running account
+                             pszPassword                // Password of the account
+  );
+  if (schService == NULL) {
+    printf("CreateService failed (%lu)\n", GetLastError());
+    CloseServiceHandle(schSCManager);
+    return -3;
+  } else
+    printf("Service installed successfully\n");
+
+  // Centralized cleanup for all allocated resources.
+  CloseServiceHandle(schService);
+  CloseServiceHandle(schSCManager);
+  return 0;
+}
+
+/*!
+ * \brief Uninstall service
+ *
+ * \details Stop and remove the service from the local service control
+ *          manager database.
+ *
+ * \param pszServiceName - the name of the service to be removed.
+ *
+ * \note If the function fails to uninstall the service, it prints the
+ *   error in the standard output stream for users to diagnose the problem.
+ *
+ * \see https://learn.microsoft.com/windows/win32/services/deleting-a-service
+ */
+int UninstallService(LPTSTR pszServiceName) {
+  SC_HANDLE schSCManager = NULL;
+  SC_HANDLE schService = NULL;
+  SERVICE_STATUS ssSvcStatus = {};
+
+  // Open the local default service control manager database
+  schSCManager = OpenSCManager(NULL, NULL, SC_MANAGER_CONNECT);
+  if (schSCManager == NULL) {
+    printf("OpenSCManager failed (%lu)\n", GetLastError());
+    return -1;
+  }
+
+  // Open the service with delete, stop, and query status permissions
+  schService = OpenService(schSCManager, pszServiceName, SERVICE_STOP | SERVICE_QUERY_STATUS | DELETE);
+  if (schService == NULL) {
+    printf("OpenService failed (%lu)\n", GetLastError());
+    CloseServiceHandle(schSCManager);
+    return -2;
+  }
+
+  // Try to stop the service
+  if (ControlService(schService, SERVICE_CONTROL_STOP, &ssSvcStatus)) {
+    _tprintf(_T("Stopping %s."), pszServiceName);
+    Sleep(1000);
+
+    while (QueryServiceStatus(schService, &ssSvcStatus)) {
+      if (ssSvcStatus.dwCurrentState == SERVICE_STOP_PENDING) {
```

Comment:
You don't have any condition where we give up waiting. So the only way this while loop will end is if the service couldn't be queried or it's not in the SERVICE_STOP_PENDING state.

Is there any value in adding a maximum timeout?

Reply:
If it isn't SERVICE_STOP_PENDING state, it can't proceed to the next step. Otherwise, the program will be uncontrollable.